lang10/contracts/auction.sympl

```diff
@@ -1,127 +1,300 @@
+            cvm.error(f"The bid amount must be larger than initial price ${auction.initial_price}.")
+    else:
+        if amount <= last_bid.amount:
+            cvm.error(f"The bid amount must be larger than current bid amount ${last_bid.amount}.")
+
+    channel = auction.channel
+    with PostTxArgs(channel):
+        _bid(id, amount)
 
-# Executable to place a bid on a product (idenfitied by the auction id)
 @executable
-def _bid(product_id: Identifier, amount: int) -> None:
+def _bid(id: Identifier, amount: int) -> None:
+    auction = get_auction(id)
     bidder: str = cvm.tx.key_alias
-    auction = cvm.storage.get(PUBLIC, AuctionStatic, product_id)
+    channel : ChannelName = cvm.tx.write_channel
+
     if isinstance(auction, None):
-        cvm.error(f"Unknown product id={product_id}")
+        cvm.error(f"You do not have access to auction {id}.")
     if auction.closed:
-        cvm.error(f"Closed auction id={product_id}")
+        cvm.error(f"Auction {id} is closed.")
+    
     last_bid = auction.last_bid
     if isinstance(last_bid, None):
         if amount < auction.initial_price:
-            cvm.error("The bid amount must be superior to initial price id={product_id}")
+            cvm.error(f"The bid amount must be larger than initial price ${auction.initial_price}.")
     else:
         if amount <= last_bid.amount:
-            cvm.error("The bid amount must be superior to previous bid amount id={product_id}")
+            cvm.error(f"The bid amount must be larger than current bid amount ${last_bid.amount}.")
+
     bid = Bid(bidder = bidder, amount = amount)
     auction.last_bid = bid
-    cvm.storage.put(product_id, auction)
-    cvm.job_complete(std.json(auction))
+    cvm.storage.put(id, auction)
+
+    c_str : str = channel
+    if c_str == 'PUBLIC':
+        cvm.create_event('Bid Placed', std.json(auction))
+    else:
+        cvm.create_event('Bid Placed', {"channel": channel, "notice": "Bid placed on silent auction."})
+
+
+#################################
+# Admin API and Implementations #
+#################################
+
+# Creates an admin.
+@clientside
+def create_admin(new_admin: KeyAlias) -> None:
+    existing_admins = cvm.storage.query(AdminStatic).values()
+
+    # The first admin creation happens automatically.
+    # Anyone created beyond that must be created by an Admin and shouldn't be an Admin themselves already.
+    if len(existing_admins) > 0:
+        if not is_admin():
+            cvm.error(f"Only an admin may perform this function.")
+            
+        na_ka : str = new_admin
+        for a in existing_admins:
+            a_ka : str = a.id
+            if a_ka == na_ka:
+                cvm.error(f"{new_admin} is already an admin.")
+
+    with PostTxArgs(PUBLIC):
+        _create_admin(new_admin)
+
+@executable
+def _create_admin(new_admin: KeyAlias) -> None:
+    existing_admins = cvm.storage.query(AdminStatic).values()
+
+    if len(existing_admins) > 0:
+        if not is_admin():
```

Comment:
Move `is_admin` to the top of the function.

Then add a `where` clause to see if the specific admin you're searching for is present.